doc/api18970.md

```diff
@@ -0,0 +1,175 @@
+**may**, **must**, or **must not** be invoked from a specific context.
+
+## Definitions
+
+### yield (thread)
+
+A thread yields when it invokes an operation that transitions it to
+*Suspended* or *Waiting*.
+
+### pre-empted (thread)
+
+A thread is pre-empted when it transitions from *Running* to *Ready*
+without having yielded.  (Only preemptible threads can be pre-empted.)
+
+### yielding (function)
+
+A function is yielding when it invokes (directly or indirectly) an
+operation that causes the current thread to yield.
+
+A function is non-yielding when it is not yielding.
+
+### thread-safe
+
+A function is thread-safe if its behavior is correct when invoked from
+multiple threads at the same time.
+
+#### Commentary
+
+Most public API will satisfy this condition; some private API may not.
+
+### reentrant
+
+A function is reentrant if its behavior is correct when it is invoked by
+(indirect) recursion from the same thread.
+
+### interrupt-safe
+
+A function is interrupt-safe if its behavior is not affected by
+concurrent access from interrupts.
+
+#### Commentary
+
+Most public API will satisfy this condition; some private API may not.
+
+We need to be able to say succinctly "Unless otherwise specified all API
+functions are interrupt-safe" and expect people to know what that means.
+A specific example would be the GPIO API.  Because GPIO write functions
+may be invoked from ISRs read-modify-write code like:
+
+```
+u32_t out = gpio->OUT;
+gpio->OUT ^= (out & ~mask) | (value & mask);
+```
+
+*must* be wrapped in a spin-lock to be interrupt-safe.  Many current
+implementations do not satisfy this requirement.
+
+On the other hand internal functions may be written to assume they are
+called with interrupts disabled, or a specific lock held.
+
+**TODO** standard marking?
+
+### atomic
+
+An operation is atomic if the steps it makes internally cannot be
+affected by nor visible to interleaving executions, such as from
+interrupts or thread pre-emption.
+
+An operation that is atomic is by definition interrupt-safe.
+
+An operation that is atomic is by definition thread-safe.
+
+### asynchronous
+
+A function is asynchronous if it may return before the operation it
+initiates has completed.  An asynchronous function must provide a
+mechanism by which completion is reported, e.g. a callback, event, or
+signal.
+
+#### Commentary
+
+Note that asynchronous is orthogonal to non-yielding.  Some API may
+provide completion information through a callback, but may yield while
+waiting.
+
+### queued
+
+A function is queued if it is asynchronous and allows multiple
+operations to be outstanding at any time.
+
+#### Commentary
+
+This concept is necessary because of operations like
+`spi_transceive_async()` which returns its result through a signal, but
+will yield if the device is already processing an asynchronous
+operation.
+
+Queued operations are rare(? do not exist?) in Zephyr, and require that
```

Comment:
No objections to the definition of the term, but I guess it goes without saying that each term should have at least one user before being included in the glossary.

Reply:
Agreed.  Marked it TBD and proposed, expanded on the reasoning behind why this is relevant, and suggested it's probably the wrong thing to do.  I expect to remove it before this gets merged.